### PR TITLE
In events module: allow choice in how the event server is implemented

### PR DESCRIPTION
**Pull request by relevitt**
This pull request addresses #468. At least two of us (@nlagaros is the other) are using this code in our respective applications. I've used and tested the twisted version for more than a year.

All updates to events.py in the SoCo master branch have been included in this pull request. It's time consuming for me to keep merging master updates into my branch, hence this pull request.

### CODE REVIEW

**Comment by ghcs27** [comment]
To document the discussion at #538 here:

 
@KennethNielsen [commented on 13. Feb 18](https://github.com/SoCo/SoCo/issues/538#issuecomment-365306608) on issue #538:
> I will comment on that PR shortly, but the short version of the reply is that I find that the parallel implementation in the same module doesn't do anything good for readability of an already complicated module and I do not like the idea of pulling twisted in as a general dependency, so I would not favor that PR.
> 
> What I could support however, is an implementation in another module, say events_twisted with a similar interface. That would allow us to keep the simpler implementation in both modules and make the twisted dependency optional, by catching import errors and formatting an easy to understand exception that twisted will need to be installed separately. The downside to this is of course that someone will need to take ownership of this module and port improvements.
> 
> In general, I would also really like it if the bugs that you have with the ordinary implementation and reported and kept updated, so that hopefully they can be fixed in the future.

 
@KennethNielsen [commented on 13. Feb 18](https://github.com/SoCo/SoCo/issues/538#issuecomment-365307714) on issue #538:
> As a side note to the comment above, you can of course borrow as much from the main events module as possible (e.g. the parse_event_xml and possibly import classes and override method with changes), to make the continued porting less of a hassle.

 
@relevitt [commented on 13. Feb 18](https://github.com/SoCo/SoCo/issues/538#issuecomment-365322676) on issue #538:
> If that's the only way we are permitted to proceed, then of course that's how we will proceed, and I'm happy to be the owner of the module.
>
> It does have the downside you mention that updates to events.py will have to be ported into events_twisted, which is the downside that prompted this effort.
>
> I'll make one further plea for the approach set out in the pull request, trying to address the points you've raised.
>
> First, the pull request does not make twisted a dependency of soco (just as the approach you advocate also doesn't do this). Nothing is ever imported from twisted, unless the application which imports the events module from soco has already imported a twisted reactor. We could easily go further and make it an express requirement for anyone wishing to use events_twisted that they first import it explicitly, so that nobody inadvertently ends up using events_twisted (i.e. they imported a twisted reactor into their app, but nevertheless wanted to use a threaded event listener).
>
> Secondly, the parallel imports you mention (events_base and events_twisted) have nothing sonos specific in them. Everything relevant to sonos remains in events.py. These parallel imports have only the bare backend for an event notification server (with a consistent interface). In one module, this is implemented with requests and threads. In the other module it is implemented with the twisted framework.

 
@KennethNielsen [commented on 14. Feb 18](https://github.com/SoCo/SoCo/issues/538#issuecomment-365542891) on issue #538:
> I see that the PR has changed a lot since the first time I saw it. I will give it another look in its current form.

**Comment by ghcs27** [comment]
One note about the current implementation: The decision whether to use twisted is a global one, and it is decided when the module is imported for the first time. Afterwards, this cannot be changed.
This is unfortunate, since you might have one script using the normal version and one using the twisted version - and you get errors if you try to run them consecutively on the same interpreter.

The only solution I see for this problem is to reverse the dependencies: make the classes in events.py abstract base classes, and the classes in events_base(_twisted).py inherit from them. Then the user could import events_base or events_base_twisted as he likes. (Of course, the module names would be different then.)

**Comment by relevitt** [comment]
@ghcs27 a helpful suggestion and, without yet looking at the detail of it, I think we can do it that way. 

I think there are three possibilities at the moment:

1) have a separate events_twisted module into which updates to events would need to be ported
2) your suggestion
3) the approach set out in this PR.

Shall I wait for more feedback on these possibilities, before implementing?

**Comment by ghcs27** [comment]
Please do wait for a minute, I just figured out that there are other options to consider.

**Comment by ghcs27** [comment]
My previous suggestion is probably rather time-consuming, and would require multiple inheritance.
I just found out that you can make multiple instances of the same module via [`imp.load_module()`](https://docs.python.org/3/library/imp.html#imp.load_module) (see [this](https://stackoverflow.com/questions/11170949/how-to-make-a-copy-of-a-python-module-at-runtime) stackoverflow question).
We could use this to make a normal and a twisted version from the same code file.

**Comment by KennethNielsen** [comment]
Hi @ghcs27 and @relevitt 

So I glanced the code and first of let me say thanks for the effort and that I think it is fair enough, that we have some version of this merged, so the work won't be in vain and that I think this code is nicely implemented along its design.

That being said, I still have some reservations about this approach. First and foremost, it seems like SoCo is being used by a lot of tinkerers with small programming experience to do some small cool thing. For that reason, I think the interface and the docs for "stock" SoCo should be kept as simple as possible. For hacking-on-SoCo reasons I also definitely prefer a solution without "mixed code".

With regards to the discussion here, I think being able to run threaded events AND async events in the same interpreter, is a (unnecessarily) steep requirement, that it would be my opinion that we wouldn't need to honor. I also think, as @relevitt already mentioned, that choosing one model or the other **must be explicit** and not automatic, so we will need an implementation that allows for that.

@ghcs27 btw, the imp solution is deprecated in newer Pythons, so we would need a Py2 and Py3 differing solutions.

I still wonder what this would look like if we turned it upside down (one file with shared code ``events_base.py`` and two frontends ``events.py`` and ``events_twisted.py``). Could I persuade you @relevitt to implement that approach in another branch (and PR) so that we might better compare.

Another reason, besides from the ones already mentioned above, that I prefer that approach, is that making model-specific changes at the top level will include some weird switching behavior. Consider this example. One of the reasons, if I remember correctly, that it is used, is that twisted handles network disconnect/and reconnect more gracefully. I have been pondering a solution and am thinking that I might make a network monitor and on lost network queue a network lost event in the subscriptions (so that the user will know that it has happened and can restart the subscriptions herself) and somehow reset subscriptions active subscriptions and make unsubscribe doesn't do anything. Implementing something like that will include having an element either be or not be there in events.py (the network monitor) and control over subscriptions. Sure this could be done, but I think it would fit better into a two-heads kind of approach, essentially because it is a fix for something that twisted does.

Another complications (that @ghcs27 ) already touched upon in the review is locks. Execution of threaded code will be interleaved by the interpreter down to a single line level, whereas code executed by the event loop/reactor will (if I understand it correctly) by synchronous within a single function/method (unless that function/methods makes use or any of the special deferreds/yields/yields from-stuff) (at least they are in regular asyncio). This somewhat complicates parallel execution correctness considerations.

What do you think @relevitt?

**Comment by relevitt** [comment]
Hi @KennethNielsen ,

Thank you for feedback. I'm happy to go with your suggestion (one file with shared code events_base.py and two frontends events.py and events_twisted.py). I'll aim for release 0.15.

Let me know if you'd like me to include something for network disconnects, with an idea how it would look and what it would do for the threaded version. 

My app checks every second if the network is reachable, looked at from the perspective of the machine on which the app is running. If the network goes down, it shuts down all soco functionality (not just the event listener). Everything is restarted when the network goes up. The app includes a sonos group object for each group on the sonos network. Each sonos group object holds information about the group. If the network goes down, the app destroys each sonos group object. New ones are created, when the network restarts.

In other words, for me the network reachable question has wider scope than events and I'm not sure I'd need a network checking solution within events. Then again, I haven't thought this through in detail yet.

My app doesn't currently check for when sonos devices drop off the network (for whatever reason), although I think @nlagaros does (it does check every second for group changes). Similar considerations would apply.

**Comment by KennethNielsen** [comment]
@relevitt thanks. Keep this PR open, in case it doesn't work out with the other solution. Might I also, for the other PR, suggest implementing the changes directly on top of master, so that we loose some of the backlog of reintegration merges.

With regards to network checking I agree that it has wider scope, but it would need to also be in threade events, because, if I remember it correctly, trying to unsubscribe when the network has been lost acts very wonky. I didn't mean to say that you necessarily needed to implement that as well ;) but I would like to see the code you use.

File: soco/events.py [hunk 1]
```diff
@@ -204,165 +264,165 @@ def __setattr__(self, name, value):
...
                 reachable by the Sonos net.
-
-        Note:
-            The port on which the event listener listens is configurable.
-            See `config.EVENT_LISTENER_PORT`
         """
 
         # Find our local network IP address which is accessible to the
         # Sonos net, see http://stackoverflow.com/q/166506
-        with self._start_lock:
```
[the remark this answers is not in the document]
Comment (relevitt) [reply soco/events.py]: @ghcs27 This is an interesting question. I've been trying to review the history. What I can see is that this lock is used only in one place, namely the start method of the EventListener class. When the events module is imported, one instance of this is created, at line 638 of the events module:

event_listener = EventListener()

The start function of event_listener is called only by the subscribe method of a Subscription object. As far as the events module is concerned, the subscribe method of a Subscription object and the event_listener.start are executed in the main thread. 

There is therefore no potential race condition here. It isn't possible to execute event_listener.start simultaneously in two different threads by importing the events module, instantiating a Subscription object and calling its subscribe method. No potential race condition = no need for a lock. It's redundant.

However, if another application imports the events module, runs the subscribe method of two or more Subscription objects in separate threads created by that application, then there will be a potential race condition and you may have each Subscription object trying to start the event_listener simultaneously.

What I'm unsure about is whether that is a problem the events module should be trying to solve. Put another way, I suspect this is but one of many examples of race conditions that could arise if another application uses threads, but which aren't addresses in the soco modules.

What do you think?

**Comment by relevitt** [comment]
@KennethNielsen @ghcs27 I've been making good progress with this, but have hit a stumbling block with the proposal to have two modules **events** and **events_twisted**, each of which would import from a module called **events_base**, with the user explicitly importing from one module or the other.

The problem is that if someone imports **soco.events_twisted** (or imports from **soco.events_twisted**) the python interpreter will, before it does anything else, run the **__init__** module in the soco directory (unless it has already done this).

The **__init__** module has the line:

`from .core import SoCo`

This sets off a chain of imports, including (in soco/services.py", line 43):

`from .events import Subscription`

In other words, any attempt to install from the **events_twisted** module, is pre-empted by **__init__**, which installs from the **events** module.

One brute force method around this is to suggest to users that they rename the **events_twisted.py** file as **events.py**, but that doesn't feel like the right solution. It would mean every upgrade of soco would have to be accompanied by an explicit renaming of files.

The approach I was adopting, of checking whether a twisted reactor is running and then automatically making the appropriate imports based on that, is another simple way around the problem. Any other thoughts or suggestions?

**Comment by ghcs27** [comment]
The real problem with the import from services.py is: Which kind of subscription should Service.subscribe() return?
For that, I'd suggest making an `EVENTS_MODULE` variable in config.py and using `config.EVENTS_MODULE.Subscription` in subscribe().

Note that with this implementation, it is possible for the modules events.py and events_twisted.py to coexist. This means that it's okay if we import events.py and the user imports events_twisted.py later.
(That's one of the most important benefits of this approach!)

**Comment by relevitt** [comment]
@ghcs27 thank you! 

I've changed that reference in `services` to:

`subscription = config.EVENT_SUBSCRIPTION`

and added the following to `config`:

```
class EventInstaller(object):

    def __new__(cls, *args, **kwargs):
        from .events import Subscription
        global EVENT_SUBSCRIPTION
        EVENT_SUBSCRIPTION = Subscription
        return Subscription(*args, **kwargs)

EVENT_SUBSCRIPTION = EventInstaller
``` 

Without any user customisation, the first time an instance of `config.EVENT_SUBSCRIPTION` is created, the instantiating of a `config.EventInstaller` object will commence. Upon instantiation of a `config.EventInstaller` object, because the `__new__` method of the class has been overridden, an instance of `events.Subscription` will be created and returned instead. Additionally, `config.EVENT_SUBSCRIPTION` will be updated to refer to `events.Subscription`, so that this installation step will occur only once.

A user can change this behaviour with:

```
import soco
from soco.events_twisted import Subscription
soco.config.EVENT_SUBSCRIPTION = Subscription
```

I think this accomplishes everything we want. A couple of notes:

- At first, I wanted to define `EventInstaller` in `events_base`, but that leads to circular imports and fails. This is why the definition appears in `config`.

- The purpose of this approach is to avoid importing both `events` and `events_twisted`. In order to do this, I think one must give the user an opportunity to import `events_twisted` before `events` has been imported. Therefore, we need to defer importing `events` until the user has had this opportunity. That's why the importing of `events` occurs in `config.EventInstaller.__new__`. I can't think of a better way of doing this.

**Comment by ghcs27** [comment]
Why is it so important to avoid importing both modules?

**Comment by relevitt** [comment]
@ghcs27 I think we could import both modules if we prefer that approach. Each module instantiates an instance of EventListener. If one imported both modules, I think one instance of EventListener would be dormant. I suppose that in general I try to avoid that sort of outcome. 

The other alternative would be to add the following to `soco.config` (instead of what's in my previous message):

```
from .events import Subscription
EVENT_SUBSCRIPTION = Subscription
```

If one then did this:

```
import soco
from soco.events_twisted import Subscription
soco.config.EVENT_SUBSCRIPTION = Subscription
```

then the outcome would be much the same for the twisted user, except that `soco.events` would also have been imported and a dormant instance of `soco.events.EventListener` would have been created. I guess that's OK.

**Comment by ghcs27** [comment]
Ah, I didn't think about the EventListener. But with normal use, they won't both run at the same time, so that shouldn't be a real issue.
Is there a reason why you prefer
`EVENT_SUBSCRIPTION = soco.events.Subscription` over
`EVENTS_MODULE = soco.events`?

**Comment by relevitt** [comment]
@ghcs27 Either approach works in Python3. But I've just found that neither approach works in Python 2.7. One once again hits the problem of circular imports:

```
:~/soco_0.15$ python -B events_twisted_test.py 
Traceback (most recent call last):
  File "events_twisted_test.py", line 6, in <module>
    import soco
  File "/home/relent/soco_0.15/soco/__init__.py", line 14, in <module>
    from .core import SoCo
  File "/home/relent/soco_0.15/soco/core.py", line 18, in <module>
    from . import config
  File "/home/relent/soco_0.15/soco/config.py", line 56, in <module>
    from . import events
  File "/home/relent/soco_0.15/soco/events.py", line 26, in <module>
    from .events_base import (
  File "/home/relent/soco_0.15/soco/events_base.py", line 15, in <module>
    from . import config
ImportError: cannot import name config
```
The use of an Installer which defers importing also avoids circular imports. I haven't found a way of setting the config variable (whether it's `config.EVENT_SUBSCRIPTION` or `config.EVENTS_MODULE`) without first importing the `soco.events` module into `soco.config`. That's where the circular import problem arises. All other approaches I've tried (i.e. without importing the `soco.events` module into `soco.config`) result in an error when attempting to set the variable in `soco.config`. For example:

```
python3 -B events_twisted_test.py 
Traceback (most recent call last):
  File "events_twisted_test.py", line 6, in <module>
    import soco
  File "/home/relent/soco_0.15/soco/__init__.py", line 14, in <module>
    from .core import SoCo
  File "/home/relent/soco_0.15/soco/core.py", line 18, in <module>
    from . import config
  File "/home/relent/soco_0.15/soco/config.py", line 71, in <module>
    EVENTS_MODULE = soco.events
AttributeError: module 'soco' has no attribute 'events'

```

**Comment by ghcs27** [comment]
What about initializing `EVENTS_MODULE` from events.py itself, like we do with `SOCO_CLASS`?

**Comment by relevitt** [comment]
To initialise EVENTS_MODULE in events, you have to import events into events. I don't know how to do that. Adding `import events` to events works in python 2.7, but not in python 3. Adding `import .events` works in python 3, but not in 2.7.

Initialising EVENT_SUBSCRIPTION in events is easy enough and doesn't raise this wrinkle, because Subscription is defined in events and no further imports are required.  

Either way, if we're initialising anything in events, then events should be imported via __init__.py, otherwise no initialisation will occur unless the user imports events. 

In 0.14, events is imported at startup only because it is imported by services. That's how our discussion began. We can continue to have services import events, simply for the purpose of ensuring initialisation occurs.

To summarise the whole proposal:

**services.py**
`import .events` (replaces `from .events import Subscription`)
`subscription = config.EVENT_SUBSCRIPTION` (replaces `subscription = Subscription`)

**config.py**
`EVENT_SUBSCRIPTION = None`

**events.py**
```
if config.EVENT_SUBSCRIPTION is None:
    config.EVENT_SUBSCRIPTION = Subscription
```
**To install events_twisted**
```
import soco
from soco.events_twisted import Subscription
soco.config.EVENT_SUBSCRIPTION = Subscription

```

**Comment by relevitt** [comment]
Alternatively, one could handle initialisation in services.py. Either of the following would work:

```
from . import events
from . import config
config.EVENTS_MODULE = events
subscription = config.EVENTS_MODULE.Subscription
```

```
from .events import Subscription
from . import config
config.EVENT_SUBSCRIPTION = Subscription
subscription = config.EVENT_SUBSCRIPTION
```

**Comment by ghcs27** [comment]
Those options all look good. Which would you favor?
You can get the current module with `sys.modules[__name__]`, so that's also an option.

**Comment by relevitt** [comment]
I've been going with:

```
from . import events
from . import config
config.EVENTS_MODULE = events
subscription = config.EVENTS_MODULE.Subscription
```
as it tracks your original suggestion and is basically the same as the way things are being done in 0.14, where services imports from events, but happy to adopt another approach. Sorry it took me such a long time to get there! Thank you for reminding me about sys.modules - I'd overlooked that.

Not too much left to do now. I'm going through your other suggestions (above), as to which I've reinstated `self._start_lock = threading.Lock()`. After that, I'll update doc strings and then do the pull request.
